ktor-server/ktor-server-plugins/ktor-server-forwarded-header/jvmAndNix/src/io/ktor/server/plugins/forwardedheaders/ForwardedHeaders.kt

```diff
@@ -0,0 +1,170 @@
+        connectionPoint: MutableOriginConnectionPoint,
+        forward: ForwardedHeaderValue?
+    ) {
+        if (forward == null) {
+            return
+        }
+
+        if (forward.proto != null) {
+            val proto: String = forward.proto
+            connectionPoint.scheme = proto
+            URLProtocol.byName[proto]?.let { p ->
+                connectionPoint.port = p.defaultPort
+            }
+        }
+
+        if (forward.forParam != null) {
+            val remoteHost = forward.forParam.split(",").first().trim()
+            if (remoteHost.isNotBlank()) {
+                connectionPoint.remoteHost = remoteHost
+            }
+        }
+
+        if (forward.host != null) {
+            val host = forward.host.substringBefore(':')
+            val port = forward.host.substringAfter(':', "")
+
+            connectionPoint.host = host
+            port.toIntOrNull()?.let { connectionPoint.port = it }
+                ?: URLProtocol.byName[connectionPoint.scheme]?.let { connectionPoint.port = it.defaultPort }
+        }
+    }
+
+}
+
+/**
+ * Parsed forwarded header value. All fields are optional as proxy could provide different fields
+ * @property host field value (optional)
+ * @property by field value (optional)
+ * @property forParam field value (optional)
+ * @property proto field value (optional)
+ * @property others contains custom field values passed by proxy
+ */
+public data class ForwardedHeaderValue(
+    val host: String?,
+    val by: String?,
+    val forParam: String?,
+    val proto: String?,
+    val others: Map<String, String>
+)
+
+/**
+ * [ForwardedHeaders] plugin allows you to obtain information headers from the original request in reverse proxy
+ * setups. For more information see https://datatracker.ietf.org/doc/html/rfc7239
+ */
+public val ForwardedHeaders: ApplicationPlugin<Application, ForwardedHeadersConfig, PluginInstance> =
+    createApplicationPlugin("ForwardedHeaders", createConfiguration = { ForwardedHeadersConfig() }) {
+
+
+        /**
+         * A key for application call attribute that is used to cache parsed header values
+         */
+        val forwardedParsedKey: AttributeKey<List<ForwardedHeaderValue>> =
+            AttributeKey("ForwardedParsedKey")
+
+
+        fun parseForwardedValue(value: HeaderValue): ForwardedHeaderValue {
+            val map = value.params.associateByTo(HashMap(), { it.name }, { it.value })
+
+            return ForwardedHeaderValue(
+                map.remove("host"),
+                map.remove("by"),
+                map.remove("for"),
+                map.remove("proto"),
+                map
+            )
+        }
+
+        fun ApplicationRequest.forwardedHeaders() =
+            headers.getAll(HttpHeaders.Forwarded)
+                ?.flatMap { it.split(',') }
+                ?.flatMap { parseHeaderValue(";$it") }?.map {
+                    parseForwardedValue(it)
+                }
+
+        onCall { call ->
```

Comment:
This callback is invoked too late. Logging, monitoring and some other plugins will be executed with unmodified value. Consider creating a hook for `Setup` phase

Reply:
Same comment for other plugin